<a href="https://colab.research.google.com/github/winsatid62/RPI-ML_Class-Fall2025/blob/main/MANE_4961_HW_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
import numpy as np


## 1. (10 pts.) Consider feature vectors, 𝐱𝟏 = ⌈2 3 4 8 9⌉$^T$ and 𝐱𝟐 = ⌈2 −3 −4 89⌉$^T$

(a) How many features are there in 𝐱𝟏 and 𝐱𝟐?

(b) What are the 𝐿1, 𝐿2, and 𝐿∞ norms for each feature vector?

In [11]:
# Code for problem 1

# Initializing the vector
# (using row vector instead of column vector as it donesn't affect norm
# calculation)
x_1 = np.array([2,3,4,8,9])
x_2 = np.array([2,-3,-4,89])

x_1_1norm = np.linalg.norm(x_1, ord=1)
x_1_2norm = np.linalg.norm(x_1, ord=2)
x_1_infNorm = np.linalg.norm(x_1, ord=np.inf)

print("Norms of x1")
print(f"L1 Norm: {x_1_1norm}")
print(f"L2 Norm: {x_1_2norm}")
print(f"Linf Norm: {x_1_infNorm}\n")

x_2_1norm = np.linalg.norm(x_2, ord=1)
x_2_2norm = np.linalg.norm(x_2, ord=2)
x_2_infNorm = np.linalg.norm(x_2, ord=np.inf)

print("Norms of x2")
print(f"L1 Norm: {x_2_1norm}")
print(f"L2 Norm: {x_2_2norm}")
print(f"Linf Norm: {x_2_infNorm}")

Norms of x1
L1 Norm: 26.0
L2 Norm: 13.19090595827292
Linf Norm: 9.0

Norms of x2
L1 Norm: 98.0
L2 Norm: 89.16277250063504
Linf Norm: 89.0


### Problem 1 Answer:
(a) There are 4 features in $\mathbf x_{1}$ and 5 features in $\mathbf x_{2}$

(b) For $\mathbf x_{1}$:
- L1 norm = 26
- L2 norm ≈ 13.191
- L∞ norm = 9

For $\mathbf x_{2}$:
- L1 norm = 98
- L2 norm ≈ 89.163
- L3 norm = 89

## 2. (20 pts.) A color image of size 1024x1024 is input to an algorithm which outputs a 64x64 color image representing some important portions of the original image.
[Hint: a color image comprises of multiple channels]

(a) If the input is converted to a vector, 𝐱 , calculate the length of the feature vector.

(b) If the output is converted to a vector, 𝐲, calculate the length of the output vector.

(c) The algorithm relates the input and output vectors, 𝐱 and 𝐲 respectively, as 𝐲 = 𝑊𝐱 + 𝐛, where W is a matrix and 𝐛 is a vector. How many elements are there in
matrix W? How many elements are in vector b?

In [ ]:
# Code for problem 2


### Problem 2 Answer:

(a) [3145728 features]

There are three channel in an RGB color image, so each pixel is represented by 3 features. There are 1024x1024 = 1048576 pixels, so the total feature count is 3145728 features

(b) [12,288 features]

Same process as (a): 64x64x3 = 12288 features

(c) [W: 38654705664 elements, b: 12288]

Matrix W map x to y, which have 3145728 features and 12288 features respectively, so it must be a 12288x3145728 matrix (Assumimg we are working with column vectors). As such, W has 38654705664 elements. Vector b need to match the dimension of y so it will have the same size as y: 12288 elements


## 3. (10 pts.) Calculate 1-norm, 2-norm, ∞-norm, and Frobenius norm of
$$W= \begin{bmatrix}
1 & -1 \\
2 & 0
\end{bmatrix} $$

### Problem 3 Answer:

## 4. (10 pts.) Use the Iris dataset and the code shown in class, but choose 3 out of the 4 original features instead of two and calculate the accuracy score of classification.

### Problem 4 Answer:

## 5. (10 pts.) In class, we used a neural network to learn how to calculate square root. Use that code and ideas to teach a neural network how to calculate the 7-th root of any integer between 1 and 100.

### Problem 5 Answer:

## 6. (40 pts.) Use the Iris dataset and the neural network code shown in class to create a neural network model that takes sepal length as input and predicts petal length as output.

[Hint: Use Numpy or Pandas to wrangle Iris dataset to help define your input and output. You are free to define the number of neurons in each layer.]

### Problem 6 Answer: